### ISSUE TITLE
Add source location info for accessor related warnings

### ISSUE THREAD

**Issue by tdavidcl** [AdaptiveCpp/AdaptiveCpp#1311]
**Describe the motivation for the feature request**
Currently those two warnings
```
[AdaptiveCpp Warning] kernel_operation: Could not find embedded pointer in kernel blob for this requirement; do you have unnecessary accessors that are unused in your kernel?
```

```
[AdaptiveCpp Warning] dag_direct_scheduler: Detected a requirement that is neither of discard access mode (SYCL 1.2.1) nor no_init property (SYCL 2020) that accesses uninitialized data. Consider changing to discard/no_init. Optimizing potential data transfers away.
```
gives no information about the location of the code triggering them

**Describe the solution you'd like**
passing something like `std::source_location`  to queue `submit` and forwarding the object to `buffer_memory_requirement::bind` and `hipsylc::rt::submit_requirement`

**If applicable, describe alternatives you've considered**
Beside passing source_location i don't think it is achievable

**Additional context**
- In large codebase tracking this warning tends to be not a straightforward task.
- I can provide a std::source_location like object that works in c++17

**Comment by nilsfriess** [comment]
Thanks for the suggestions.

I did some quick testing and I also think this is definitely something that can be implemented using `std::source_location` (or something similar in C++17).
For this code, for instance
```c++
// test.cc
#include <sycl/sycl.hpp>

int main() {
  sycl::queue q;

  sycl::buffer<int> b{1024};

  q.submit([&](auto &h) {
    auto a = b.get_access<sycl::access::mode::read_write>(h);

    h.single_task([=](){});
  });
}
```
I now get a warning like this
```
[AdaptiveCpp Warning] kernel_operation (originally called from test.cc:11): Could not find embedded pointer in kernel blob for this requirement; do you have unnecessary accessors that are unused in your kernel?
````
Since this is something that should probably only be enabled in debug builds or when some `-DACPP_WARN_LOCATIONS` flag is passed, it will get a bit ugly in some places.

But as you're not the first person to suggest something like this, I'd be happy to continue with this. @illuhad what do you think?

**Comment by illuhad** [comment]
I agree that this can be useful. But I'm concerned about bloating up the DAG node. Currently, a lot there is implemented using smart pointers, but the long term plan is to make them directly store their data for performance (similarly to the change we've made to the execution hints). In that case source location could potentially increase the size of every DAG node.

So if we are doing this, I think we should find ways to ensure that this really only affects the runtime when it is actually needed. E.g. USM operations should not have to haul it around, and as you suggest perhaps there could be a special flag of build configuration.

**Comment by nilsfriess** [comment]
In my quick test I did something like this in `handler::single_task`
```c++
  template <typename KernelName = __hipsycl_unnamed_kernel, typename KernelType>
#ifndef NDEBUG
  void single_task(KernelType kernelFunc,
                   nostd::source_location loc = nostd::source_location::current()) {
#else
  void single_task(KernelType kernelFunc) {
#endif
    this->submit_kernel<KernelName, rt::kernel_type::single_task>(
      NDEBUG_ARG(loc)
      sycl::id<1>{0}, sycl::range<1>{1}, sycl::range<1>{1}, kernelFunc);
  }
```
where `NDEBUG_ARG(arg)` becomes `arg,` if `NDEBUG` is not set and is just empty otherwise. 
As I said, this is quite ugly but in builds with `NDEBUG` set, there is no overhead. If we actually implement something like this, there should probably be some flag to _enable_ this explicitly, i.e., only add the source_location argument if the flag is set.

**Comment by tdavidcl** [comment]
Yes i agree this is only for convenience when developing, it does not really have a need in release builds.

On the other side, since you forward only the filename + line of the call, it would be "only" a `const char*`, would that be significant compared to the size of a DAG node ?

Also maybe to have a similar syntax between debug and release builds you could to something like 

```c++
struct debug_source_location{
.....

#ifndef NDEBUG
  std::string _file;
  int _line;
#endif
}
```
which remove the need to change the signature in the calls by just changing the underlying object

**Comment by illuhad** [comment]
> On the other side, since you forward only the filename + line of the call, it would be "only" a const char*, would that be significant compared to the size of a DAG node ?

Even if it is just `const char*`, things still pile up. If `std::string` is involved, it will be costly due to non-trivial copy constructor and dynamic memory allocation.

I wonder, if we say that we only need the functionality for debug builds - couldn't you also just set a breakpoint?

**Comment by tdavidcl** [comment]
Yes this was the workaround i did previously, however it's **very** inconvenient when it's SYCL coupled with MPI, you would need to setup multiple debuggers using a mpirun, ... Like it's doable clearly but i was thinking that this may be useful tooling for acpp, especially for newcomers that may not know where to set such breakpoint ...

**Comment by tdavidcl** [comment]
Hi, 
I'm willing to help on this one, @nilsfriess do you have a finished version ? How may I help here ?

**Comment by nilsfriess** [comment]
Sorry, completely forgot to reply.

We had some internal discussion about this, I don't remember exactly now but I think we came up with the following compromise:
- Simply remove the `dag_direct_scheduler: Detected a requirement that is neither...` warning
- Print the kernel name stored in the `kernel_operation` that emits the `Could not find embedded pointer...` warning. If the kernel is a lambda, it might still be a bit difficult to exactly identify it because the name will look something like this `ZZ4mainENK3$_0clERN7hipsycl4sycl7handlerEEUlvE_` but if you kernel is defined via an `operator()` of some class, then you will at least get the class name.

We thought that passing the source location down from every possible code path might be difficult. What do you think? Do you think printing the kernel name is sufficient?

**Comment by tdavidcl** [comment]
Yes for sure passing a source location might be a significant change, especially if it has to be linked to a dag node or something similar…

I would be fine with just the kernel name being printed (of course source location would be better but maybe too annoying on your side), the issue currently being the fact that beside debugger magic there is now ways to track the issue. Having the kernel name is better than nothing :)

**Comment by illuhad** [comment]
> Print the kernel name stored in the kernel_operation that emits the Could not find embedded pointer... warning. If the kernel is a lambda, it might still be a bit difficult to exactly identify it because the name will look something like this ZZ4mainENK3$_0clERN7hipsycl4sycl7handlerEEUlvE_ but if you kernel is defined via an operator() of some class, then you will at least get the class name.

Note that the kernel name stored in `kernel_operation` is totally unused starting with the branch `feature/rewrite-kernel-cache` and is thus probably going the way of the dinosaurs very soon. So it's not something we should rely on for a long term solution of the problem.
